### ISSUE TITLE
NAL Unit exceeds subsample definition

### ISSUE THREAD

**Issue by matthuisman** [peak3d/inputstream.adaptive#208]
Hi @peak3d 

3 out of 4 streams work with this provider.

But one stream errors with a bunch of 'NAL Unit exceeds subsample definition'.
I have tried the stream via their official Android app (exoplayer) and it does playback (using same MPD file & Widevine license).

Using KODI RC1 Windows 64bit

Let me know if you need a live test stream and I can sort for you

**Comment by blairhandley** [comment]
Hey Matt having trouble this morning non of the channels will work for me on your kodi add-on am using a mobile fanpass

**Comment by matthuisman** [comment]
@blairhandley
Please start a new issue on my repo for the plugin:
https://github.com/matthuisman/plugin.video.fanpass
(Note: Make sure your now using KODI 18 and using latest version of the add-on)

**Comment by matthuisman** [comment]
Looks like this NAL Unit error is happening on all streams on Android.

A user supplied log below (it has "banned" add-ons in it)
https://paste.kodi.tv/ixugigotuf

**Comment by matthuisman** [comment]
@peak3d 
Seems this is popping up with some other add-ons now as well.

Any ideas?
Did you want me to get you a test stream?

**Comment by liberty-developer** [comment]
@matthuisman can you pls. download the m4s file (printed in log)?

**Comment by joshjryan** [comment]
Additional info, for reference: when I see this same error (referenced in https://github.com/maxgalbu/xbmc.plugin.video.nba/issues/87), I hear audio, but no video is displayed.
If it's helpful, I can try to also get additional logging sometime today.

**Comment by matthuisman** [comment]
@peak3d 

No working stream, just video = Not work (log: https://pastebin.com/raw/cnyNXjeD)
No working stream, just audio = Not work (log: https://pastebin.com/raw/j93jg9R5)

The error with just audio is different though.
DecryptSampleData: Decrypt failed with error: 2 and key: 5EA7B1992BE5BA5F7200BA05F97F0000

Working stream, just audio = Works (log: https://pastebin.com/raw/LtX8mLpJ)
This has the same decrypt failed error but still works??

Here is also the capture of the request the app makes
![untitled](https://user-images.githubusercontent.com/6225961/49470368-b9143380-f86e-11e8-9655-da5efee293bd.png)

I will try email you a 2 links for testing :)
1st will be this stream that won't play on any device.
2nd will be a stream that won't play on Android.

**Comment by cablegoon** [comment]
I am having the same issue on a fire tv and sheild (android tv) with various releases on kodi. Let me know if you want logs from my end too. 
this happens with the NBA league pass app

**Comment by micahg** [comment]
I get this too with my sportsnet addon. It works on Ubuntu (64bit) but fails on my Android machines. Please let me know if I can assist.

**Comment by joshjryan** [comment]
Note: I am still seeing the same issue with RC3.

**Comment by Lakersfan16** [comment]
@cablegoon @joshjryan Any solutions for the only Audio problem yet?

**Comment by UniqueSports** [comment]
We have the same **audio only** issue on Android devices (checked on a mobile and an NVIDIA SHIELD). Our source is same as the NBA league pass app. Can provide logs and the video chunks in question if interested.

@matthuisman
You might wanna remove sensitive info such as tokens from the logs before posting them publicly.

**Comment by joshjryan** [comment]
If helpful, here are the logs from when I receive the "NAL Unit exceeds subsample definition" error: http://paste.kodi.tv/ijeqojuduy

**Comment by matthuisman** [comment]
Looks like @peak3d is M.I.A at the moment?

No activity on his Github since November.
Hopefully he's just taking a well earned holiday and is back soon :)

**Comment by lyossy** [comment]
hey guys
for those of you who are able to compile kodi , I've discovered a patch that seems to eliminate this issue .
I have not tested it extensively , please test and inform the results .
The patch is in the file that outputs the message ( the one that ends in "_jni.cpp" ) and is inserted just before the test the causes the message .  
here it is :
// bypass all validity tests if packet is complete
     if (packet_in == packet_in_e )//packet completed ok
	{
	data_out.SetDataSize(data_out.GetDataSize() + data_in.GetDataSize() + configSize + (4 - fragInfo.nal_length_size_) * nalunitcount);
	return AP4_SUCCESS;
	}
 
// old test , do not change this in the original code , just insert the above lines before this test .
// as i've modified the old test for debugging , no need to modify original code .
        if (nalsize + fragInfo.nal_length_size_ + nalunitsum > *Clear + *Cipher)

**Comment by lyossy** [comment]
this file is in the wvdecryper folder .
If anyone succeeds in building an apk with this patch and it works , please share it .

**Comment by UniqueSports** [comment]
> 
> 
> It did

@lyossy
Hey  Nice work! 

Do you know how to cross compile the patch on Ubuntu for Android arm devices ?

**Comment by lyossy** [comment]
kodi themselves have a file called "readme.md" that has instructions compiling on ubuntu 16.04 for all 
3 platforms : arm7,arm8,x86 ..
google "compiling kodi for android" .
I was not able to generate the full  apk , as on my setup it reached 99% and failed .
I was able to build just the buggy library for arm v7 .
It is called "libssd._uv.so" .
by replacing this library with the one i built with said patch , i was able to watch streams that previously
only played audio .

**Comment by UniqueSports** [comment]
@lyossy 
Hey, thanks for the reply. Yes, I was using the guide on https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md 

I'm not sure how to generate the apk from the modified local version. I'm not a C++ guy and have even less experience with the kodi development itself. 

Did you just target arm7 when building the library itself (not the apk) ?

**Comment by lyossy** [comment]
first of all i tried to build the apk . this sets up the enviionrment for compiling the addons .
later , i just ran the make process in the library folder .
I'm not sure if it's allowed , but i can send you a patched apk that i created that includes  the patched library .Can somebody comment if this is allowed ?
If so , send me a link to a location i can upload to ( file is 62 MB )

**Comment by lyossy** [comment]
works on mibox 3 ( amlogic s905x androidtv )

**Comment by Lakersfan16** [comment]
@lyossy what exactly do I need to change now to get video and audio? (Sorry im new to this)

**Comment by lyossy** [comment]
are you a programmer or at least 
proficient in programming ?

**Comment by Lakersfan16** [comment]
@lyossy im a bit proficient

**Comment by lyossy** [comment]
you need to edit a cpp file whose name
ends with "_jni.cpp" in the wvdecryper
folder . you have to insert the 3 lines of
cpp code i posted a few hours ago just
before the test that causes the "NAL"
error to appear .
than , rebuild the kodi apk ( make sure
you have rebuilt the addons ) . 
uninstall your current leia installation 
and install the apk you built .

**Comment by Lakersfan16** [comment]
@lyossy you said earlier you can send us the apk. So is this still the case?

**Comment by Lakersfan16** [comment]
I think as long as you dont say that it is an official version it should be alright. 
https://uploadfiles.io

**Comment by lyossy** [comment]
Hi
Is this a public location ? I prefer sending just to you .

**Comment by Lakersfan16** [comment]
@lyossy no its not. You just have to message me the Link that you get after uploading.

**Comment by lyossy** [comment]
What is your email ?

**Comment by Lakersfan16** [comment]
justin.pullmann@gmx.de

**Comment by lyossy** [comment]
OK , I sent you the link . note that the file will be available for an hour , as I am not sure I'm allowed to share it . 
Remember you need to uninstall previous Leia installation , otherwise apk will not install ( it is not accepted as an update )
Please inform the forum how it goes .

**Comment by lyossy** [comment]
You can test it immediately using archived games . also I noted that the league pass add-on is not working well after installation of the add-on . after a reboot it works much better

**Comment by Lakersfan16** [comment]
It works perfectly. Thanks man. Finally i can watch the NBA on the big screen. Appreciate your effort. 🐐

**Comment by lyossy** [comment]
Have fun 😋
Can you specify your device , CPU , android version .
Verified on MINIX neo u9 , a logic s912 , regular android 6.0

**Comment by Lakersfan16** [comment]
@lyossy Device: Orbsmart S84 
CPU: AML S912 Octa Core ARM Cortex-A53
Android 7.1

**Comment by peak3d** [comment]
Sorry for absense, was anyone able to pull a sample stream (segment) ?
Simply look into kodi log, and pick the URL of one downloaded segment (video I guess)

The "fix" provided above will not be accepted until the root cause is clear.

Thx for the work done here!

**Comment by MRLB** [comment]
@peak3d

Would it be possible to merge this code approach with your code to get a running Android Version for the NBA League Pass. Thanks :-) I am looking forward to watch NBA on Kodi again

**Comment by blairhandley** [comment]
Hey there could I please have a copy aswell

**Comment by lyossy** [comment]
Sorry , I am not allowed to distribute it  . however , I believe an official solution is coming soon . if you can't wait , you can build the apk yourself . please note that this is just an ad hoc patch , until an official solution is available , and is proven working only on current NBA international league pass . I have no idea if it works with other sources . I did this out of a personal need , and made it public to help fellow lp subscribers . 
It was not my intention to offend anybody .

**Comment by UniqueSports** [comment]
@peak3d 
Here are few chunks form a while back.  https://www.mediafire.com/folder/vbtb4z8u84v1y/ this contain a playlist, two init chunks (video) from two different bitrates, one regular video chunk and an audio init chunk.

let me know if you need anything else.

**Comment by peak3d** [comment]
Thx @UniqueSports , will look today at the issue

**Comment by matthuisman** [comment]
@peak3d 
I've also just emailed you some files :)

**Comment by peak3d** [comment]
I pushed the fix, version is 2.3.13, pls. verify, thx!

**Comment by dwane3** [comment]
Where is download 2.3.13?

**Comment by dwane3** [comment]
peak3d, please update to 2.3.13 version on official repo..

**Comment by matthuisman** [comment]
Tested on Windows and confirm the 4th stream that wasn't working is now working.
Will try on Android soon

**Comment by matthuisman** [comment]
Hmm..

Android seems to be playing but is just white noise and KODI's GUI goes all "funny" as well.

Using KODI 18 RC 5.2 ARMV8
And used your inputstream android-aarch64/inputstream.adaptive-2.3.13.zip

Also tried other Widevine contact and it played fine.

Screenshot:
![screenshot_20190123-121226_kodi](https://user-images.githubusercontent.com/6225961/51571980-cbbef680-1f08-11e9-85ee-5c3209cd8e1e.png)

**Comment by joshjryan** [comment]
Latest nightly fixed it for me.
Android
32-bit
FireTV stick 4k
NBA League Pass addon (@ivankokan's branch, which handles the PIL issues from the official repo)

**Comment by matthuisman** [comment]
Fixed in v2.3.13 https://github.com/peak3d/inputstream.adaptive/commit/22baed345295ef4e6fe103fa48dff847bf7d91cd

**Comment by micahg** [comment]
Hi, I can also confirm this fixed my issues. Tested 2.3.13 on my firestick and mi box.

Thanks very much!

**Comment by cablegoon** [comment]
works for me - thanks for everyones help

**Comment by matthuisman** [comment]
I have found that I get no audio on Windows KODI 18 on that same troublesome stream.
KODI 18 Android works fine though.
![screenshot000](https://user-images.githubusercontent.com/6225961/52031118-cf452400-257f-11e9-8c09-4ae319a0e428.png)

To get a better debug log, I switched Inputstream to Audio only.
Played one of the good streams and then played the stream that fails.

I will email @peak3d the needed files to debug :)
